core/metacling/src/TCling.cxx

```diff
@@ -2324,6 +2364,27 @@ void TCling::RegisterModule(const char* modulename,
    }
 }
 
+void TCling::AddAvailableIndentifiers(TSeqCollection& Idents) {
+   clang::CompilerInstance& CI = *GetInterpreterImpl()->getCI();
+   ASTContext &C = CI.getASTContext();
+
+   // Do not do anything if we have no global module index.
+   // FIXME: This is mostly to real with false positives in the TTabCom
+   // interface for non-modules.
+   if (!fCxxModulesEnabled)
+      return;
+
+   if (IdentifierInfoLookup *External = C.Idents.getExternalIdentifierLookup()) {
```

Comment:
Weird - I was convinced that `IdentifierInfo` is always an (unscoped) name, i.e. `A::BB::CCC` would contain `A`, `BB`, `CCC`. Do you have a test that's sensitive to this, e.g. that we don't offer `CCC` as completion on the global scope?

Reply:
That is correct. We control what gets registered as identifier [here](https://github.com/root-project/root/blob/master/core/metacling/src/TCling.cxx#L1112). Let's take your example `A::BB:CCC` being present in module `M`. In code where we have `namespace A{ namespace BB{}}; using namespace A::BB; CCC* c = new CCC();` we need the GMI to store CCC->M mapping.

This means that it will give CCC as an identifier for tabcompletion. And yes, we have such a test -- that's the TH1DModel in the RDF namespace.

We do not have a reasonable way to improve tab completion in this case for the moment.

Reply:
@vgvassilev I am confused.  In the case of ```A::BB:CCC``` **without** any using statement.  Would this code means that ```CC<tab>``` will auto-complete to CCC or fail (as expected).   What about ```A::BB::CC<tab>```?

Reply:
It will autocomplete in both cases.

Reply:
In the case where you have:
```
namespace A { class DDD; };
namespace B { namespace C { class DDD; } };
```
what is the output on screen of
```
root [] DD<tab>
```

Reply:
Nothing -- that's not in the GMI so that's not in the completion set.

Reply:
Ok.  Then I am confused.  The previous comments asked:
```
in the case of A::BB:CCC without any using statement. Would this code means that CC<tab> will auto-complete to CCC or fail (as expected). What about A::BB::CC<tab>?
```
and the answer was
```
It will autocomplete in both cases.
```
From that I understood that with
```
namespace A { namespace B { class CCC; } };
```
would lead to
```
CC<tab>
autocompletes to
CCC
```
and
```
A::BB:CC<tab>
autocompletes to
A::BB::CCC
```
but that seems to be contradicted by the last answer (very similar setup):

> Nothing -- that's not in the GMI so that's not in the completion set.

So what am I missing?

Reply:
Sorry, the comments was assuming that A::BB::CCC is part of a dictionary (has a module for it) and went into the global module index. An example for this is ROOT::RDF::TH1DModel. There if you type TH1<tab> the results will contain TH1DModel although it may not be reachable (as we may not have had using `namespace ROOT::RDF`). The artificially created examples, such as A::BB::CCC -- on CC<tab> won't complete as CCC was not in the GMI.

Reply:
Ah but I meant to say "what if those types where in a dictionary and hence in the GMI ?".
I.e. if the types:
```
namespace A { namespace B { class CCC; }
namespace A { class DDD; };
namespace B { namespace C { class DDD; } };
```
are in a dictionary and hence in the GMI what does:
```
CC<tab>
DD<tab>
B::C::DD<tab>
```
prints?

Reply:
```root.exe -l -b
root [0] TH1
TH1
TH1C
TH1D
TH1DModel
TH1Editor
TH1F
TH1I
TH1K
TH1S
```
```root.exe -l -b
root [0] ROOT::RDF::
MakeCsvDataFrame
MakeSqliteDataFrame
MakeTrivialDataFrame
RCsvDS
RInterface<ROOT
RTrivialDS
TH1DModel
TH2DModel
TH3DModel
TProfile1DModel
TProfile2DModel
```

Reply:
Well this is a little bit "annoying" :( ... What is the source of the problem? (I.e. is it the issue @Axel-Naumann is pointing out about line 2377?)

Reply:
Yes, the same issue.

Reply:
Fair enough.  For the record the current behavior is reversely "bad" as even with the `using namespace ROOT::RDF `, it does not list `TH1DModel`.